pkg/registry/service/ipallocator/etcd/etcd.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	etcderr "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors/etcd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/ipallocator"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+)
+
+// Etcd exposes a service.Allocator that is backed by etcd.
+// TODO: allow multiple allocations to be tried at once
+// TODO: subdivide the keyspace to reduce conflicts
+// TODO: investigate issuing a CAS without reading first
+type Etcd struct {
+	lock sync.Mutex
+
+	alloc  ipallocator.Snapshottable
+	helper tools.EtcdHelper
+	last   string
+}
+
+// Etcd implements ipallocator.Interface and service.IPRegistry
+var _ ipallocator.Interface = &Etcd{}
+var _ service.IPRegistry = &Etcd{}
+
+const baseKey = "/ranges/serviceips"
```

Comment:
Yeah I'll probably take the hit there too because i have to do UID allocation in OpenShift (which isn't IP based).  So we'll have 3 use cases which should be enough to get the right mix.

----- Original Message -----
> > +// TODO: allow multiple allocations to be tried at once
> > +// TODO: subdivide the keyspace to reduce conflicts
> > +// TODO: investigate issuing a CAS without reading first
> > +type Etcd struct {
> > +	lock sync.Mutex
> > +
> > +	alloc  ipallocator.Snapshottable
> > +	helper tools.EtcdHelper
> > +	last   string
> > +}
> > +
> > +// Etcd implements ipallocator.Interface and service.IPRegistry
> > +var _ ipallocator.Interface = &Etcd{}
> > +var _ service.IPRegistry = &Etcd{}
> > +
> > +const baseKey = "/ranges/serviceips"
> 
> Good enough.  When we get to a port allocator, we can finish whatever is
> needed to make a second range config here.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29901815